haskell/private/cc_wrapper.py.tpl

```diff
@@ -0,0 +1,812 @@
+            # reference to the list of arguments to forward. The handler must
+            # return True if it consumes the argument, and return False if
+            # another handler should consume the argument.
+            if self._handle_output(arg, args, out):
+                pass
+            elif self._handle_library(arg, args, out):
+                pass
+            elif self._handle_library_path(arg, args, out):
+                pass
+            elif self._handle_linker_arg(arg, args, out):
+                pass
+            elif self._handle_print_file_name(arg, args, out):
+                pass
+            elif self._handle_compile(arg, args, out):
+                pass
+            else:
+                yield arg
+
+            for out_arg in out:
+                yield out_arg
+
+    def _handle_output(self, arg, args, out):
+        if arg == "-o":
+            # Remember the output filename.
+            self.output = next(args)
+            out.extend(["-o", self.output])
+            return True
+        else:
+            return False
+
+    def _handle_library(self, arg, args, out):
+        if arg == "-l" or arg == "--library":
+            library = next(args)
+        elif arg.startswith("-l"):
+            library = arg[2:]
+        elif arg.startswith("--library="):
+            library = arg[len("--library="):]
+        else:
+            return False
+
+        # Remember the required libraries.
+        self.libraries.append(library)
+        out.append("-l{}".format(library))
+
+        return True
+
+    def _handle_library_path(self, arg, args, out):
+        if arg == "-L" or arg == "--library-path":
+            library_path = next(args)
+        elif arg.startswith("-L"):
+            library_path = arg[2:]
+        elif arg.startswith("--library-path="):
+            library_path = arg[len("--library-path="):]
+        else:
+            return False
+
+        # Shorten the library search paths. On Windows library search paths may
+        # exceed the maximum path length.
+        shortened = shorten_path(library_path)
+        # Remember the library search paths.
+        self.library_paths.append(shortened)
+        out.append("-L{}".format(shortened))
+
+        return True
+
+    def _handle_linker_arg(self, arg, args, out):
+        if arg == "-Xlinker":
+            ld_arg = next(args)
+            if self._prev_ld_arg is None:
+                if ld_arg == "-rpath":
+                    self._prev_ld_arg = ld_arg
+                else:
+                    out.extend(["-Xlinker", ld_arg])
+            elif self._prev_ld_arg == "-rpath":
+                self._prev_ld_arg = None
+                self._handle_rpath(ld_arg, out)
+            else:
+                # This indicates a programmer error and should not happen.
+                raise RuntimeError("Unhandled _prev_ld_arg '{}'.".format(self._prev_ld_arg))
+            return True
```

Comment:
From the [gcc manpage](http://man7.org/linux/man-pages/man1/gcc.1.html#OPTIONS), `-Xlinker` description: 

```
 When using the GNU linker, it is usually more convenient to pass
           arguments to linker options using the option=value syntax than as
           separate arguments.  For example, you can specify -Xlinker
           -Map=output.map rather than -Xlinker -Map -Xlinker output.map.
           Other linkers may not support this syntax for command-line
           options.
```

Do we need to handle that case?

Reply:
Done. AFAIK e.g. BSD ld does not accept that syntax and GHC doesn't emit it, but for completeness it makes sense to handle it.